R/decorated-future.R

```diff
@@ -0,0 +1,27 @@
+decorated_future <- function(inner, pre = function() {}, post = function() {}) {
+  structure(
+    list(
+      inner = inner,
+      pre = pre,
+      post = post
+    ),
+    class = c("DecoratedFuture", class(inner))
+  )
+}
+
+#' @export
+run.DecoratedFuture <- function(future, ...) {
+  future$pre()
+  future::run(future$inner)
+}
+
+#' @export
+resolve.DecoratedFuture <- function(x, ...) {
+  future::resolve(x$inner)
+  x$post()
+}
+
+#' @export
+resolved.DecoratedFuture <- function(x, ...) {
```

Comment:
Do we need this special `resolved()` method? I had assumed we would automatically inherit `resolved()` from the inner future.

Reply:
We're not inheriting, we're wrapping. This means that we need to write this (and perhaps other) forwarding code.

Reply:
Ah, because a decorated `future` is a list containing a conventional `future`. Yes, my mistake.